From 55f7c2ae7bc608105231554c93de73fd6283cd4e Mon Sep 17 00:00:00 2001
From: Cedric Roux <cedric.roux@eurecom.fr>
Date: Tue, 12 May 2026 11:09:18 +0200
Subject: [PATCH 19/21] NR-DC: get NR-DC parameters from configuration file

All configured NR-DC combinations are searched, the first one that
is possible is chosen. A combination is possible if the MCG cell is
the one where the UE is connected and the SCG cell exists (a DU
using the SCG band is connected to the CU).

Signed-off-by: Cedric Roux <cedric.roux@eurecom.fr>
---
 openair2/RRC/NR/rrc_gNB_nrdc.c | 28 +++++++++++++++++++++++++---
 1 file changed, 25 insertions(+), 3 deletions(-)

diff --git a/openair2/RRC/NR/rrc_gNB_nrdc.c b/openair2/RRC/NR/rrc_gNB_nrdc.c
index 2c4ed606b8..dcdcbf18c2 100644
--- a/openair2/RRC/NR/rrc_gNB_nrdc.c
+++ b/openair2/RRC/NR/rrc_gNB_nrdc.c
@@ -373,11 +373,33 @@ void rrc_gnb_nrdc_start(gNB_RRC_INST *rrc, gNB_RRC_UE_t *ue)
   if (ue->nrdc)
     return;
 
-  /* look for an NR-DC combination */
-  int mcg_band = 77;
-  int scg_band = 261;
+  /* look for an NR-DC combination, configured and available */
+  nr_rrc_cell_container_t *ue_pcell = rrc_get_pcell_for_ue(rrc, ue);
+  if (ue_pcell == NULL) {
+    LOG_W(NR_RRC, "NR-DC: UE %d has no PCell, don't try to activate NR-DC\n", ue->rrc_ue_id);
+    return;
+  }
+
+  int mcg_band = -1;
+  int scg_band = -1;
+  for (int i = 0; i < rrc->nrdc_config.combination_count; i++) {
+    nr_rrc_cell_container_t *mcg_cell = get_cell_by_band(&rrc->cells, rrc->nrdc_config.combinations[i].mcg_band);
+    nr_rrc_cell_container_t *scg_cell = get_cell_by_band(&rrc->cells, rrc->nrdc_config.combinations[i].scg_band);
+    /* MCG cell must be the one used by the UE */
+    if (mcg_cell == ue_pcell && scg_cell != NULL) {
+      mcg_band = rrc->nrdc_config.combinations[i].mcg_band;
+      scg_band = rrc->nrdc_config.combinations[i].scg_band;
+    }
+  }
+
+  if (mcg_band == -1) {
+    LOG_W(NR_RRC, "NR-DC: UE %d: no NR-DC combination found, NR-DC will not be activated\n", ue->rrc_ue_id);
+    return;
+  }
 
   /* a combination is found, start NR-DC, ask for UE capabilities */
+  LOG_I(NR_RRC, "NR-DC: UE %d: trying band combination MCG band %d and SCG band %d\n", ue->rrc_ue_id, mcg_band, scg_band);
+
   /* allocate the NR-DC state data in the UE */
   nrdc_ue_state_t *nrdc = calloc_or_fail(1, sizeof(*nrdc));
   ue->nrdc = nrdc;
-- 
2.48.1

